File: assignments/haskell/robot-simulator/robot-simulator_test.hs
import Test.HUnit (Assertion, (@=?), runTestTT, Test(..), Counts(..))
import System.Exit (ExitCode(..), exitWith)
import Robot (Bearing(..), Robot, mkRobot,
              coordinates, simulate,
              bearing, turnRight, turnLeft)

exitProperly :: IO Counts -> IO ()
exitProperly m = do
  counts <- m
  exitWith $ if failures counts /= 0 || errors counts /= 0 then ExitFailure 1 else ExitSuccess

testCase :: String -> Assertion -> Test
testCase label assertion = TestLabel label (TestCase assertion)

main :: IO ()
main = exitProperly $ runTestTT $ TestList
  [ TestList robotTests ]

robotTests :: [Test]
robotTests =
  [ testCase "turning edge cases" $ do
    North @=? turnRight West
    West @=? turnLeft North
  , testCase "robbie" $ do
    let robbie = mkRobot East (-2, 1)
    East @=? bearing robbie
    (-2, 1) @=? coordinates robbie
    let movedRobbie = simulate robbie "RLAALAL"
    West @=? bearing movedRobbie
    (0, 2) @=? coordinates movedRobbie
    mkRobot West (0, 2) @=? movedRobbie
  , testCase "clutz" $ do
    let clutz = mkRobot North (0, 0)
    mkRobot West (-4, 1) @=? simulate clutz "LAAARALA"
  , testCase "sphero" $ do
    let sphero = mkRobot East (2, -7)
    mkRobot South (-3, -8) @=? simulate sphero "RRAAAAALA"
  , testCase "roomba" $ do
    let roomba = mkRobot South (8, 4)
    mkRobot North (11, 5) @=? simulate roomba "LAAARRRALLLL"
  ]

File: assignments/clojure/robot-simulator/robot_simulator_test.clj
(ns robot_simulator.test (:require [clojure.test :refer :all]))
(load-file "robot.clj")

(def robbie (robot {:x -2 :y 1} :east))
(deftest can-get-vals
  (is (= :east (:bearing robbie)))
  (is (= {:x -2 :y 1} (:coordinates robbie))))

(deftest can-turn
  (is (= :north (turn-right :west)))
  (is (= :west  (turn-left :north))))

(deftest can-simulate
  (is (= :west (:bearing (simulate "RLAALAL" robbie))))
  (is (= {:x 0 :y 2} (:coordinates (simulate "RLAALAL" robbie)))))

(def clutz  (robot {:x 0 :y  0} :north))
(def sphero (robot {:x 2 :y -7} :east))
(def roomba (robot {:x 8 :y  4} :south))

(deftest simulate-clutz
  (let [clutz (simulate "LAAARALA" clutz)]
    (is (= :west (:bearing clutz)))
    (is (= {:x -4 :y 1} (:coordinates clutz)))))

(deftest simulate-sphero
  (let [sphero (simulate "RRAAAAALA" sphero)]
    (is (= :south (:bearing sphero)))
    (is (= {:x -3 :y -8} (:coordinates sphero)))))

(deftest simulate-roomba
  (let [roomba (simulate "LAAARRRALLLL" roomba)]
    (is (= :north (:bearing roomba)))
    (is (= {:x 11 :y 5} (:coordinates roomba)))))

(run-tests)

